pkg/controller/controller.go

```diff
@@ -0,0 +1,283 @@
+type Pair struct {
+	Key  string
+	Text string
+}
+
+// Content defines the displayable parts of the token retrieval form.
+type Content struct {
+	UserLabel     string
+	CategoryLabel string
+	ReasonLabel   string
+	TTLLabel      string
+	Categories    []Pair
+	TTLs          []Pair
+}
+
+// FormDetails represents all the input and content used for the token retrievlal form.
+type FormDetails struct {
+	WindowName  string
+	Origin      string
+	PageTitle   string
+	Description string
+	Content     Content
+	Category    string
+	Reason      string
+	TTL         string
+	Errors      map[string]string
+}
+
+// SuccessDetails represents the data used for the success page and the postMessage response to the client.
+type SuccessDetails struct {
+	WindowName  string
+	Origin      string
+	PageTitle   string
+	Description string
+	Token       string
+}
+
+// ErrorDetails represents the data used for the 400 page.
+type ErrorDetails struct {
+	PageTitle   string
+	Description string
+	Message     string
+}
+
+var (
+	categories = []string{"explanation", "breakglass"}
+	ttls       = []string{"15", "30", "60", "120", "240"}
+)
+
+func New(h *render.Renderer) *Controller {
+	return &Controller{
+		h: h,
+	}
+}
+
+func (c *Controller) HandlePopup(allowList []string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		formDetails := getFormDetails(r)
+
+		// Initial page load, just render the page
+		if r.Method == http.MethodGet {
+			// set some defaults for the form
+			formDetails.Category = categories[0]
+			formDetails.TTL = ttls[0]
+
+			c.h.RenderHTML(w, "popup.html.tmpl", formDetails)
+			return
+		}
+
+		// Form submission
+		if r.Method == http.MethodPost {
+			// 1. Check if the origin is part of the allowlist
+			origin := r.FormValue("origin")
+			if validOrigin, err := validateOrigin(origin, allowList); err != nil || !validOrigin {
+				var m string
+				if err != nil {
+					m = err.Error()
+				} else if !validOrigin {
+					m = "Unexpected origin provided"
+				}
+
+				t := http.StatusText(http.StatusBadRequest)
+				forbiddenDetails := &ErrorDetails{
+					PageTitle:   t,
+					Description: t,
+					Message:     m,
+				}
+				c.h.RenderHTMLStatus(w, http.StatusBadRequest, "400.html.tmpl", forbiddenDetails)
+				return
+			}
+
+			// 2. Validate input
+			if !validateForm(formDetails) {
+				c.h.RenderHTML(w, "popup.html.tmpl", formDetails)
+				return
+			}
+
+			// 3. [TODO] Request a token
+			token := "token_from_server"
+
+			// 4. Redirect to a confirmation page with context, ultimately needed to postMessage back to the client
+			successDetails := &SuccessDetails{
+				PageTitle:   "JVS - Successful token retrieval",
+				Description: "Successful token page",
+				Token:       token,
+				Origin:      formDetails.Origin,
+				WindowName:  formDetails.WindowName,
+			}
+			c.h.RenderHTML(w, "success.html.tmpl", successDetails)
+		}
+	})
```

Comment:
As written, if I send an HTTP PUT or DELETE, this method does nothing. It can be a follow up, but here's how I'd structure this:

```go
func (c *Controller) HandlePopup(allowList []string) http.Handler {
  return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    switch r.Method {
      case http.MethodGet:
        c.handlePopupPost(w, r)
      case http.MethodPost:
        c.handlePopupPost(w, r)
      default:
        http.Error(w, http.MethodNotAllowed)
    }
  })
}
```